src/StreamingDataFrames/tests/test_dataframes/test_dataframe/test_dataframe.py

```diff
@@ -273,3 +275,73 @@ def test_to_topic_no_producer_assigned(self, dataframe_factory, row_factory):
                     topic=topic.name, key=b"test_key", value={"x": "1", "y": "2"}
                 )
             )
+
+
+class TestDataframeStateful:
+    def test_apply_stateful(self, dataframe_factory, state_manager, row_factory):
+        topic = Topic("test")
+
+        def stateful_func(value, state: State):
+            current_max = state.get("max")
```

Comment:
It wasn't until this example that I think I understand what you meant by prefixing the key, lol.

Am I correct in understanding this is how it works structurally:

- folder: consumer_group
  - folder: topic_T
    - folder: partition_N
      - RDB file: "default"
        - kafka msg key: `key_X`
          - stored key: `max` (via `key_X-max`) = `10`
        - kafka msg key `key_Y`
          - stored key `max` (via `key_Y-max`) = `5`
      - RDB file: "rolling10s"
        - kafka msg key: `key_X`
          - mean?

This could be my own biases here based on how we implemented state with Fluvii, but I feel like I'd prefer to just store state without the need to define keys with it since you can store whatever object you want. The only upside I can see is if you only need a certain part of the state, you can potentially load only what you need...but you could also have way more keys to counteract that.

We can discuss it more if my understanding is correct.

I'll wait to confirm my understanding is correct though.

Reply:
Yes, you're correct.
- The actual DB dir is `<consumer_group>/<topic>/<partition>/<store_name>`
- The actual keys inside the store are prefixed with the message key. 

> The only upside I can see is if you only need a certain part of the state, you can potentially load only what you need...but you could also have way more keys to counteract that.

There's another upside - you can update only what you need. 
If you store the whole object, on each update you need to read and update the whole object too.

Reply:
Also true, yeah!

Not sure which is better overall...it's probably just a preference, or depends on how many of these keys you store. I suppose nothing is stopping you from storing a big json or something if you really needed to (would still have to specify a key and loads/dumps but whatev).

I think as long as the documentation is clear in how it works, it probably doesn't matter.

Reply:
I think the approach with individual keys gives more flexibility, especially when you need to update the state multiple times during a single function.
Also, it simplifies managing the state when you have multiple stateful functions in the pipeline.